### PR TITLE
feat(ingest/sql): column logic + join extraction

### PR DESCRIPTION
**Pull request by hsheth2**
Adds a couple new features to our SQL parser
- Best-effort column logic extraction. In case a column goes through multiple layers of transformation, we only take the final one.
  - For each column, we also annotate whether it was a direct copy or if it underwent any transformation - see the `is_direct_copy` bool.
- Join extraction. For each join clause, we try to figure out (1) the `on` clause used and (2) the full list of tables/columns involved in the join. The latter is more reliable when there's CTEs.

The column logic extraction is also now integrated into the `SqlParsingAggregator`, so our main warehouse integrations should benefit from it. The join logic extraction is not yet used anywhere.

All tests are updated, and I also added a couple new ones for tricky edge cases (e.g. self-joins).

### CODE REVIEW

**Review by sgomezvillamor — APPROVED** [review]
awesome!

beyond self-join, what's the coverage for other join variants such as inner/outer/cross joins? they could be explicitely tested

should user-facing doc be updated with the new features?

**Comment by hsheth2** [comment]
@sgomezvillamor this should work for all join types, including ones that use subqueries or CTEs

There's no user-facing portion of this yet, so I don't want to add any user-facing docs around this until it shows up in the product somewhere

File: metadata-ingestion/src/datahub/sql_parsing/sql_parsing_aggregator.py [hunk 1]
```diff
@@ -1383,7 +1390,16 @@ def _gen_lineage_for_downstream(
                             if self.can_generate_query(query_id)
                             else None
                         ),
-                        confidenceScore=queries_map[query_id].confidence_score,
+                        confidenceScore=query.confidence_score,
+                        transformOperation=(
```
Comment (treff7es) [inline metadata-ingestion/src/datahub/sql_parsing/sql_parsing_aggregator.py:1394]: Is there a reason we format as a string and not convert it into a JSON object to make it parsable by machine?

File: metadata-ingestion/tests/integration/sql_server/golden_files/procedures/DemoData.Foo.NewProc.json [hunk 2]
```diff
@@ -26,6 +26,7 @@
                     "downstreams": [
                         "urn:li:schemaField:(urn:li:dataset:(urn:li:dataPlatform:mssql,demodata.foo.age_dist,PROD),age)"
                     ],
+                    "transformOperation": "COPY: [persons].[age] AS [age]",
```
Comment (treff7es) [inline metadata-ingestion/tests/integration/sql_server/golden_files/procedures/DemoData.Foo.NewProc.json:29]: Is this correct with the square brackets?

**Review by treff7es — APPROVED** [review]

File: metadata-ingestion/tests/integration/sql_server/golden_files/procedures/DemoData.Foo.NewProc.json (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/tests/integration/sql_server/golden_files/procedures/DemoData.Foo.NewProc.json:29]: yes - mssql uses square brackets for quotes

File: metadata-ingestion/src/datahub/sql_parsing/sql_parsing_aggregator.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/sql_parsing/sql_parsing_aggregator.py:1394]: I believe this is shown in the UI as-is
